### PR TITLE
[feature] add support for clients editing statuses and fetching status revision history

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
# Description

- adds `PUT /api/v1/statuses/:id` method to support local users editing their own statuses
- updates the existing status history endpoint to return revision history for statuses (where possible)
- adds a slew of new tests for the local status edit endpoint

WIP:
- finish commenting code, especially the tests (and possibly refactor to share a bunch of the status / edit validation)
- possibly add documentation?

## Checklist

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have not leveraged AI to create the proposed changes.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [ ] I/we have commented the added code, particularly in hard-to-understand areas.
- [ ] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [ ] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/federation/dereferencing/dereferencer.go [hunk 1]
```diff
@@ -66,15 +66,15 @@ var (
 	// causing loads of dereferencing calls.
 	Fresh = util.Ptr(FreshnessWindow(5 * time.Minute))
 
-	// 10 seconds.
+	// 5 seconds.
```
Comment (tsmethurst) [inline internal/federation/dereferencing/dereferencer.go:69]: Spicy!

File: internal/federation/dereferencing/status.go [hunk 2]
```diff
@@ -1000,12 +1001,21 @@ func (d *Dereferencer) fetchStatusEmojis(
...
 
-	// Iterate over and set changed emoji IDs.
+	// Extract IDs from latest slice of emojis.
 	status.EmojiIDs = make([]string, len(emojis))
 	for i, emoji := range emojis {
 		status.EmojiIDs[i] = emoji.ID
 	}
 
+	// Combine both old and new emojis, as statuses.emojis
+	// keeps track of emojis for both old and current edits.
```
Comment (tsmethurst) [inline internal/federation/dereferencing/status.go:1011]: Ah interesting, didn't know this! Nice one.

File: internal/processing/workers/fromfediapi.go [hunk 3]
```diff
@@ -917,17 +917,25 @@ func (p *fediAPI) UpdateStatus(ctx context.Context, fMsg *messages.FromFediAPI)
 		return gtserror.Newf("cannot cast %T -> *gtsmodel.Status", fMsg.GTSModel)
 	}
 
+	var freshness *dereferencing.FreshnessWindow
+
 	// Cast the updated ActivityPub statusable object .
 	apStatus, _ := fMsg.APObject.(ap.Statusable)
 
+	if apStatus != nil {
```
Comment (tsmethurst) [inline internal/processing/workers/fromfediapi.go:925]: Ahh right.

**Comment by tsmethurst** [comment]
Bloody good work here! Honestly I think you should squerge it whenever you're happy with it, I've got no notes.

File: internal/api/model/attachment.go [hunk 4]
```diff
@@ -39,16 +42,38 @@ type AttachmentRequest struct {
...
 }
 
+// AttachmentAttributesRequest models an edit request for attachment attributes.
+//
+// swagger:ignore
+type AttachmentAttributesRequest struct {
+
+	// The ID of the attachment.
+	// example: 01FC31DZT1AYWDZ8XTCRWRBYRK
+	ID string `json:"id"`
```
Comment (tsmethurst) [inline internal/api/model/attachment.go:64]: Does this also need a `form` tag?

Comment (NyaaaWhatsUpDoc) [reply internal/api/model/attachment.go]: indeed! good catch!

**Pull request merged**